File: exponentiation_by_squaring/ExponentiationBySquaring.java
/* Algorithm to calculate base raised to power power very efficiently
*  Can be used as a great alternative to normal pow function
*  base - base of the expression
*  power - power of expression
*/

public class ExponentiationBySquaring {
    public static double exponentiationBySquaring(double base, int power) {
        if (power < 0) {	// Negative powers
            return exponentiationBySquaring(1.0 / base, -power);
        } else if (power == 0) {	// Base case 1
            return 1;
        } else if (power == 1) {	// Base case 2
            return base;
        } else if (power % 2 == 0){	// Even powers
            return exponentiationBySquaring(base * base, power / 2);
        } else {	// Odd powers
            return base * exponentiationBySquaring(base * base, (power - 1) / 2);
        }
    }

    public static void main(String[] args) {
        double base = 2;
        int power = -1;
        System.out.println(base + " raised to " + power + " is: " + exponentiationBySquaring(base, power));
    }
}

File: exponentiation_by_squaring/exponentiation_by_squaring.go
package main

import "fmt"

// ExponentiationBySquaring calculates base^power efficiently.
func ExponentiationBySquaring(base int, power int) int {
	if power < 0 { // Negative power case
		return ExponentiationBySquaring((1 / base), -power)
	} else if power == 0 { // Base case1
		return 1
	} else if power == 1 { // Base case2
		return base
	} else if power%2 == 0 {
		return ExponentiationBySquaring(base*base, power/2)
	} else {
		return base * ExponentiationBySquaring(base*base, (power-1)/2)
	}
}

func main() {
	base, power := 5, 10
	fmt.Println(ExponentiationBySquaring(base, power))
}
